multigeiger/ble.cpp

```diff
@@ -0,0 +1,138 @@
+#include <BLEServer.h>
+#include <BLEUtils.h>
+#include <BLE2902.h>
+
+#define BLE_SERVICE_HEART_RATE    BLEUUID((uint16_t)0x180D)  // 16 bit UUID of Heart Rate Service
+#define BLE_CHAR_HR_MEASUREMENT   BLEUUID((uint16_t)0x2A37)  // 16 bit UUID of Heart Rate Measurement Characteristic
+#define BLE_CHAR_HR_POSITION      BLEUUID((uint16_t)0x2A38)  // 16 bit UUID of Heart Rate Sensor Position Characteristic
+#define BLE_CHAR_HR_CONTROLPOINT  BLEUUID((uint16_t)0x2A39)  // 16 bit UUID of Heart Rate Control Point Characteristic
+#define BLE_DESCR_UUID            BLEUUID((uint16_t)0x2901)  // 16 bit UUID of BLE Descriptor
+
+BLECharacteristic bleCharHRM(BLE_CHAR_HR_MEASUREMENT, BLECharacteristic::PROPERTY_NOTIFY);
+BLEDescriptor bleDescriptorHRM(BLE_DESCR_UUID);
+BLECharacteristic bleCharHRCP(BLE_CHAR_HR_CONTROLPOINT, BLECharacteristic::PROPERTY_WRITE);
+BLEDescriptor bleDescriptorHRCP(BLE_DESCR_UUID);
+BLECharacteristic bleCharHRPOS(BLE_CHAR_HR_POSITION, BLECharacteristic::PROPERTY_READ);
+BLEDescriptor bleDescriptorHRPOS(BLE_DESCR_UUID);
+
+static bool ble_enabled = false;
+static bool device_connected = false;
+
+uint8_t flags_HRS = 0b00001001; // bit 0 --> 1 = HR (cpm) as UINT16, bit 3 --> add UINT16 Energy Expended (rolling update counter)
+uint8_t txBuffer_HRM[5]; // transmit buffer, byte[0] = flags, [1, 2] = cpm, [3, 4] = rolling update counter
+uint8_t txBuffer_HRPOS[1] = {TUBE_TYPE};
+unsigned int status_HRCP = 0;
+unsigned int cpm_update_counter = 0;
+unsigned int cpm_value = 0;
+
+bool is_ble_connected() {
+  return ble_enabled && device_connected;
+}
+
+class MyServerCallbacks: public BLEServerCallbacks {
+  void onConnect(BLEServer *pServer, esp_ble_gatts_cb_param_t *param) {
+    char remoteAddress[18];
+    sprintf(
+      remoteAddress,
+      "%02X:%02X:%02X:%02X:%02X:%02X",
+      param->connect.remote_bda[0],
+      param->connect.remote_bda[1],
+      param->connect.remote_bda[2],
+      param->connect.remote_bda[3],
+      param->connect.remote_bda[4],
+      param->connect.remote_bda[5]
+    );
+    log(INFO, "BLE device connected, remote MAC: %s", remoteAddress);
+    device_connected = true;
+  };
+
+  void onDisconnect(BLEServer *pServer) {
+    device_connected = false;
+    log(INFO, "BLE device disconnected");
+  }
+};
+
+// Callback allowing for Control Point Characteristic to reset "energy expenditure" (packet counter)
+class MyCharacteristicCallbacks: public BLECharacteristicCallbacks {
+  void onWrite(BLECharacteristic *pCharacteristic) {
+    std::string rxValue = pCharacteristic->getValue();
+    status_HRCP = (rxValue.length() > 0) ? rxValue[0] : 0;
+  }
+};
+
+void update_bledata(unsigned int cpm) {
+  if (!ble_enabled)
+    return;
+  cpm_update_counter++;
+  cpm_update_counter = cpm_update_counter & 0xFFFF;
+  if (status_HRCP > 0) {
+    log(DEBUG, "HR Control Point Value received: %d, resetting packet counter", status_HRCP);
+    cpm_update_counter = 0;
+    status_HRCP = 0;
+  }
+  cpm_value = cpm;
```

Comment:
why is that line needed? why is the global `cpm_value` needed?

Reply:
Currently it is not needed. Will remove it